recipes/m4/all/conanfile.py

```diff
@@ -14,117 +105,106 @@ class M4Conan(ConanFile):
-            conf_args.extend(["gl_cv_func_printf_directive_n=no", "gl_cv_func_snprintf_directive_n=no", "gl_cv_func_snprintf_directive_n=no"])
+            at.configure_args.extend(["gl_cv_func_printf_directive_n=no", "gl_cv_func_snprintf_directive_n=no", "gl_cv_func_snprintf_directive_n=no"])
+
             if self.settings.build_type in ("Debug", "RelWithDebInfo"):
-                autotools.link_flags.append("-PDB")
-        elif self.settings.compiler == "clang":
-            if tools.Version(self.version) < "1.4.19":
-                autotools.flags.extend(["-rtlib=compiler-rt", "-Wno-unused-command-line-argument"])
-        if self.settings.os == 'Windows':
-            conf_args.extend(["ac_cv_func__set_invalid_parameter_handler=yes"])
-
-        autotools.configure(args=conf_args, configure_dir=self._source_subfolder, build=build_canonical_name, host=host_canonical_name)
-        return autotools
-
-    @contextmanager
-    def _build_context(self):
-        env = {"PATH": [os.path.abspath(self._source_subfolder)]}
-        if self._is_msvc:
-            with tools.vcvars(self.settings):
-                env.update({
-                    "AR": "{}/build-aux/ar-lib lib".format(tools.unix_path(self._source_subfolder)),
-                    "CC": "cl -nologo",
-                    "CXX": "cl -nologo",
-                    "LD": "link",
-                    "NM": "dumpbin -symbols",
-                    "OBJDUMP": ":",
-                    "RANLIB": ":",
-                    "STRIP": ":",
-                })
-                with tools.environment_append(env):
-                    yield
-        else:
-            with tools.environment_append(env):
-                yield
-
-    def _patch_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+                at.ldflags.append("-PDB")
+        elif self.settings.compiler == "clang" and Version(self.version) < "1.4.19":
+            at.cflags.extend(["-rtlib=compiler-rt", "-Wno-unused-command-line-argument"])
+            at.cxxflags.extend(["-rtlib=compiler-rt", "-Wno-unused-command-line-argument"])
+
+        if self.settings.os == "Windows":
+            at.configure_args.extend(["ac_cv_func__set_invalid_parameter_handler=yes"])
+
+        # Needs to be called and set last, otherwise the previous arguments and flags can't be set
+        env = at.environment()
+        if is_msvc(self):
+            env.define("AR", f"{unix_path(self, self.source_folder)}/build-aux/ar-lib lib")
+            env.define("LD", "link")  # otherwise configure reports "ld" as its linker. See conan-io/conan#11922
+            env.define("NM", "dumpbin -symbols")
+            env.define("OBJDUMP", ":")
+            env.define("RANLIB", ":")
+            env.define("STRIP", ":")
+        at.generate(env)
+
+    @lru_cache(1)
```

Comment:
If I understand correctly we are caching the autotools at the build step. This way when retrieving the cache in the packaging cache we get the autotools generetate from a the state where it was in the build step?

In my opinion this doesn't make the data flow very clear. I prefer to set a `self._autotools` property in the build step that is read in the packaging step; it is less fancy but the flow is more clear.

Reply:
On the other hand, this makes it possible to perform the packaging step independent from the build step which is nice for debugging.